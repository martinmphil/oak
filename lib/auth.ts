import * as path from "path";
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { UserPool, AccountRecovery } from "aws-cdk-lib/aws-cognito";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { lambdaCommonProps } from "./fixtures/lambdaCommonProps";

export interface ICognitoOakProps {
  tableNameOak: string;
}

export class CognitoOak extends Construct {
  public readonly userPool;
  public readonly userPoolClient;
  public readonly cognitoPostConfirm;

  constructor(scope: Construct, id: string, props: ICognitoOakProps) {
    super(scope, id);

    const cognitoPostConfirm = new NodejsFunction(
      this,
      "CognitoPostConfirmOak",
      {
        ...lambdaCommonProps,
        entry: path.join(__dirname, `/handlers/cognito-post-confirm/index.ts`),
        environment: {
          DATABASE_NAME_OAK: props.tableNameOak,
        },
      }
    );
    new LogGroup(this, "LogGroupCognitoPostConfirmOak", {
      logGroupName: `/aws/lambda/${cognitoPostConfirm.functionName}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    this.cognitoPostConfirm = cognitoPostConfirm;

    const userPool = new UserPool(this, "UserPoolOak", {
      selfSignUpEnabled: true,
      passwordPolicy: {
        requireLowercase: false,
        requireUppercase: false,
        requireDigits: false,
        requireSymbols: false,
      },
      signInAliases: { email: true },
      signInCaseSensitive: false,
      accountRecovery: AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      lambdaTriggers: { postConfirmation: cognitoPostConfirm },
    });
    const domain = userPool.addDomain("CognitoDomainOak", {
      cognitoDomain: { domainPrefix: "greenstem-oak" },
    });

    const userPoolClient = userPool.addClient("UserPoolClientOak", {
      oAuth: {
        flows: {
          implicitCodeGrant: true,
        },
        callbackUrls: [
          "http://localhost:6100/can",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
          "https://www.greenstem.uk/can",
        ],
        logoutUrls: [
          "http://localhost:6100/logout.html",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/logout.html",
          "https://www.greenstem.uk/logout.html",
        ],
      },
      preventUserExistenceErrors: true,
      accessTokenValidity: cdk.Duration.minutes(1200),
      idTokenValidity: cdk.Duration.minutes(1200),
    });

    this.userPool = userPool;
    this.userPoolClient = userPoolClient;

    const signInUrl = domain.signInUrl(userPoolClient, {
      redirectUri: "https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
    });
    new cdk.CfnOutput(this, "CognitoSignInUrl", {
      value: signInUrl,
      description: "cognito sign in url",
      exportName: "signInUrl",
    });

    //
  }
}
