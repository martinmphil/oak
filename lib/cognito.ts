import { UserPool, AccountRecovery } from "aws-cdk-lib/aws-cognito";
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

export class CognitoOak extends Construct {
  public readonly userPool;
  public readonly userPoolClient;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const userPool = new UserPool(this, "UserPoolOak", {
      userPoolName: "canOak",
      selfSignUpEnabled: true,
      passwordPolicy: {
        requireLowercase: false,
        requireUppercase: false,
        requireDigits: false,
        requireSymbols: false,
      },
      signInAliases: { email: true },
      signInCaseSensitive: false,
      accountRecovery: AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    const domain = userPool.addDomain("CognitoDomainOak", {
      cognitoDomain: { domainPrefix: "greenstem-oak" },
    });

    const userPoolClient = userPool.addClient("UserPoolClientOak", {
      oAuth: {
        flows: {
          implicitCodeGrant: true,
        },
        callbackUrls: [
          "http://localhost:6100/can",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
          "https://www.greenstem.uk/can",
        ],
        logoutUrls: [
          "http://localhost:6100/logout.html",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/logout.html",
          "https://www.greenstem.uk/logout.html",
        ],
      },
      preventUserExistenceErrors: true,
      accessTokenValidity: cdk.Duration.minutes(1200),
      idTokenValidity: cdk.Duration.minutes(1200),
    });

    this.userPool = userPool;
    this.userPoolClient = userPoolClient;

    const signInUrl = domain.signInUrl(userPoolClient, {
      redirectUri: "https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
    });
    new cdk.CfnOutput(this, "cognito sign-in url", {
      value: signInUrl,
      description: "cognito sign in url",
      exportName: "signInUrl",
    });

    //
  }
}
