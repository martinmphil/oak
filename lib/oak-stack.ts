import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as path from "path";
import * as logs from "aws-cdk-lib/aws-logs";
import {
  HttpApi,
  HttpMethod,
  CorsHttpMethod,
} from "@aws-cdk/aws-apigatewayv2-alpha";
import { HttpLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha";

export class OakStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // lambda
    const canEmailAddrOak = new NodejsFunction(this, "canEmailAddrOak", {
      runtime: lambda.Runtime.NODEJS_16_X,
      architecture: lambda.Architecture.ARM_64,
      handler: "main",
      entry: path.join(__dirname, `/../src/can-email-addr-oak/index.ts`),
      bundling: { externalModules: ["aws-sdk"] },
    });

    new logs.LogGroup(this, "logGroupCanEmailAddrOak", {
      logGroupName: `/aws/lambda/${canEmailAddrOak.functionName}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const canEmailAddrIntegration = new HttpLambdaIntegration(
      "canEmailAddrIntegrationOak",
      canEmailAddrOak
    );

    // gateway
    const gatewayOak = new HttpApi(this, "gatewayOak", {
      // defaultAuthorizer: oakAuthorizor,
      corsPreflight: {
        allowHeaders: ["Authorization"],
        allowOrigins: [
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/",
          "http://localhost:6100",
        ],
        allowMethods: [CorsHttpMethod.GET],
      },
    });
    gatewayOak.addRoutes({
      integration: canEmailAddrIntegration,
      methods: [HttpMethod.GET],
      path: "/email-addr",
    });
  }
}
