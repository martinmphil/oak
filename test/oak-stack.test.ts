import * as cdk from "aws-cdk-lib";
import { Template, Match } from "aws-cdk-lib/assertions";
import * as Oak from "../lib/oak-stack";

describe("oak stack", () => {
  const app = new cdk.App();
  const stack = new Oak.OakStack(app, "MyTestStack");
  const template = Template.fromStack(stack);

  it("reproduces snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

  it("synthesises a cloud-formation template", () => {
    template.templateMatches(Match.anyValue());
  });
});
