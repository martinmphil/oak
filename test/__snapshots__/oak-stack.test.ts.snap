// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oak stack reproduces snapshot 1`] = `
Object {
  "Outputs": Object {
    "GatewayOakCognitoOakCognitoSignInUrlA7197B8E": Object {
      "Description": "cognito sign in url",
      "Export": Object {
        "Name": "signInUrl",
      },
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "GatewayOakCognitoOakUserPoolOakCognitoDomainOakF2187020",
            },
            ".auth.",
            Object {
              "Ref": "AWS::Region",
            },
            ".amazoncognito.com/login?client_id=",
            Object {
              "Ref": "GatewayOakCognitoOakUserPoolOakUserPoolClientOak44A94B7E",
            },
            "&response_type=token&redirect_uri=https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e845402ce43b66fc6f20df4a239f20f8662eb6c7f920b94cf6542dd0e64ce0f7.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DatabaseOakTableOak1FEE3FAA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "DatabaseOakdynamoSeedDataB522F3B3": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "DatabaseOakdynamoSeedDataCustomResourcePolicy3D6AA58C",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"DynamoDB\\",\\"action\\":\\"batchWriteItem\\",\\"parameters\\":{\\"RequestItems\\":{\\"",
              Object {
                "Ref": "DatabaseOakTableOak1FEE3FAA",
              },
              "\\":[{\\"PutRequest\\":{\\"Item\\":{\\"pk\\":{\\"S\\":\\"standardCatalog\\"},\\"sk\\":{\\"S\\":\\"standardCatalog\\"},\\"entityType\\":{\\"S\\":\\"catalog\\"},\\"obj\\":{\\"M\\":{\\"catalog\\":{\\"L\\":[{\\"S\\":\\"exam1\\"},{\\"S\\":\\"exam2\\"}]}}}}}},{\\"PutRequest\\":{\\"Item\\":{\\"pk\\":{\\"S\\":\\"exam1\\"},\\"sk\\":{\\"S\\":\\"exam1\\"},\\"entityType\\":{\\"S\\":\\"examData\\"},\\"obj\\":{\\"M\\":{\\"name\\":{\\"S\\":\\"Sustainable energy\\"},\\"qList\\":{\\"L\\":[{\\"S\\":\\"qText1\\"},{\\"S\\":\\"qText101\\"},{\\"S\\":\\"qText102\\"},{\\"S\\":\\"qText99\\"}]}}}}}},{\\"PutRequest\\":{\\"Item\\":{\\"pk\\":{\\"S\\":\\"exam2\\"},\\"sk\\":{\\"S\\":\\"exam2\\"},\\"entityType\\":{\\"S\\":\\"examData\\"},\\"obj\\":{\\"M\\":{\\"name\\":{\\"S\\":\\"Basic science\\"},\\"qList\\":{\\"L\\":[{\\"S\\":\\"qText1\\"},{\\"S\\":\\"qText201\\"},{\\"S\\":\\"qText202\\"},{\\"S\\":\\"qText99\\"}]}}}}}}]}},\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Ref": "DatabaseOakTableOak1FEE3FAA",
              },
              " seeded}\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "DatabaseOakdynamoSeedDataCustomResourcePolicy3D6AA58C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DatabaseOakTableOak1FEE3FAA",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DatabaseOakdynamoSeedDataCustomResourcePolicy3D6AA58C",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GatewayOakCognitoOakUserPoolOakA55CC282": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_email",
              "Priority": 1,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": false,
        },
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": Object {
          "PasswordPolicy": Object {
            "MinimumLength": 8,
            "RequireLowercase": false,
            "RequireNumbers": false,
            "RequireSymbols": false,
            "RequireUppercase": false,
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UsernameAttributes": Array [
          "email",
        ],
        "UsernameConfiguration": Object {
          "CaseSensitive": false,
        },
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "GatewayOakCognitoOakUserPoolOakCognitoDomainOakF2187020": Object {
      "Properties": Object {
        "Domain": "greenstem-oak",
        "UserPoolId": Object {
          "Ref": "GatewayOakCognitoOakUserPoolOakA55CC282",
        },
      },
      "Type": "AWS::Cognito::UserPoolDomain",
    },
    "GatewayOakCognitoOakUserPoolOakUserPoolClientOak44A94B7E": Object {
      "Properties": Object {
        "AccessTokenValidity": 1200,
        "AllowedOAuthFlows": Array [
          "implicit",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": Array [
          "http://localhost:6100/can",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/can",
          "https://www.greenstem.uk/can",
        ],
        "IdTokenValidity": 1200,
        "LogoutURLs": Array [
          "http://localhost:6100/logout.html",
          "https://main.d3dq4xzxmmo3wf.amplifyapp.com/logout.html",
          "https://www.greenstem.uk/logout.html",
        ],
        "PreventUserExistenceErrors": "ENABLED",
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "TokenValidityUnits": Object {
          "AccessToken": "minutes",
          "IdToken": "minutes",
        },
        "UserPoolId": Object {
          "Ref": "GatewayOakCognitoOakUserPoolOakA55CC282",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "GatewayOakDefaultStageD11DF7A7": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "GatewayOakF33261A6": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowHeaders": Array [
            "Authorization",
          ],
          "AllowMethods": Array [
            "GET",
            "PUT",
          ],
          "AllowOrigins": Array [
            "https://main.d3dq4xzxmmo3wf.amplifyapp.com/",
            "http://localhost:6100",
          ],
        },
        "Name": "GatewayOak",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "GatewayOakGETcandidateemail95C81B50": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "GatewayOakdefaultAuthorizer37125998",
        },
        "RouteKey": "GET /candidate-email",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "GatewayOakGETcandidateemailCandidateEmailIntegrationOak995EF089",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "GatewayOakGETcandidateemailCandidateEmailIntegrationOak995EF089": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcandidateEmailOakB10FAA83",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "GatewayOakGETcandidateemailCandidateEmailIntegrationOakPermissionCFC06881": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcandidateEmailOakB10FAA83",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GatewayOakF33261A6",
              },
              "/*/*/candidate-email",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GatewayOakGETcatalogA586C6FE": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "GatewayOakdefaultAuthorizer37125998",
        },
        "RouteKey": "GET /catalog",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "GatewayOakGETcatalogCatalogIntegrationOak71DAABED",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "GatewayOakGETcatalogCatalogIntegrationOak71DAABED": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcatalogOakB5A9DC5A",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "GatewayOakGETcatalogCatalogIntegrationOakPermissionC1E5A9AD": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcatalogOakB5A9DC5A",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GatewayOakF33261A6",
              },
              "/*/*/catalog",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GatewayOakLambdasOakcandidateEmailOakB10FAA83": Object {
      "DependsOn": Array [
        "GatewayOakLambdasOakcandidateEmailOakServiceRole49D37DD1",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "d0384fd7c0a96de8a8ef4ced08dd5afda440429b4cb9b5e93330670794a1c373.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcandidateEmailOakServiceRole49D37DD1",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GatewayOakLambdasOakcandidateEmailOakServiceRole49D37DD1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GatewayOakLambdasOakcatalogOakB5A9DC5A": Object {
      "DependsOn": Array [
        "GatewayOakLambdasOakcatalogOakServiceRoleA1239FDF",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "56a7143abf9ad31ca633fe4ca6bcf86e9dd5c66cb7058e750a2fc08622764298.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakcatalogOakServiceRoleA1239FDF",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GatewayOakLambdasOakcatalogOakServiceRoleA1239FDF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GatewayOakLambdasOakexamOak5C55EBD6": Object {
      "DependsOn": Array [
        "GatewayOakLambdasOakexamOakServiceRole1C7DDD50",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "18589ccb2d191dff88fe08791732daf5adc809638d86adb568711de7593f3596.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakexamOakServiceRole1C7DDD50",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GatewayOakLambdasOakexamOakServiceRole1C7DDD50": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GatewayOakLogGroupCandidateEmailOak29928669": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "GatewayOakLambdasOakcandidateEmailOakB10FAA83",
              },
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "GatewayOakLogGroupCatalogOak76CEB3C5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "GatewayOakLambdasOakcatalogOakB5A9DC5A",
              },
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "GatewayOakLogGroupExamOakF8497159": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "GatewayOakLambdasOakexamOak5C55EBD6",
              },
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "GatewayOakPUTexamexamId5C4E1265": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "AuthorizationType": "JWT",
        "AuthorizerId": Object {
          "Ref": "GatewayOakdefaultAuthorizer37125998",
        },
        "RouteKey": "PUT /exam/{examId}",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "GatewayOakPUTexamexamIdExamIntegrationOak8AA734D8",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "GatewayOakPUTexamexamIdExamIntegrationOak8AA734D8": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakexamOak5C55EBD6",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "GatewayOakPUTexamexamIdExamIntegrationOakPermissionD151DD6A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GatewayOakLambdasOakexamOak5C55EBD6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GatewayOakF33261A6",
              },
              "/*/*/exam/{examId}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GatewayOakdefaultAuthorizer37125998": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GatewayOakF33261A6",
        },
        "AuthorizerType": "JWT",
        "IdentitySource": Array [
          "$request.header.Authorization",
        ],
        "JwtConfiguration": Object {
          "Audience": Array [
            Object {
              "Ref": "GatewayOakCognitoOakUserPoolOakUserPoolClientOak44A94B7E",
            },
          ],
          "Issuer": Object {
            "Fn::Join": Array [
              "",
              Array [
                "https://cognito-idp.",
                Object {
                  "Ref": "AWS::Region",
                },
                ".amazonaws.com/",
                Object {
                  "Ref": "GatewayOakCognitoOakUserPoolOakA55CC282",
                },
              ],
            ],
          },
        },
        "Name": "defaultAuthorizer",
      },
      "Type": "AWS::ApiGatewayV2::Authorizer",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
